### PR TITLE
8312072: Deprecate for removal the -Xnoagent option

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to deprecate `-Xnoagent` option of the `java` command? This addresses https://bugs.openjdk.org/browse/JDK-8312072.

As noted in the JBS issue, this option currently is a no-op and provides no functionality. The commit in this PR logs a deprecation warning if this option is used when launching `java`.

Additionally, this PR also cleans up one such usage that I found in the JDK code.

`tier1`, `tier2` and `tier3` tests passed with this change. A CSR has been proposed for this deprecation https://bugs.openjdk.org/browse/JDK-8312073.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8312073](https://bugs.openjdk.org/browse/JDK-8312073) to be approved

### Issues
 * [JDK-8312072](https://bugs.openjdk.org/browse/JDK-8312072): Deprecate for removal the -Xnoagent option (**Bug** - P4)
 * [JDK-8312073](https://bugs.openjdk.org/browse/JDK-8312073): Deprecate for removal the -Xnoagent option (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [71df3079](https://git.openjdk.org/jdk/pull/14882/files/71df30790eaf6ddff184d32e7be2d1864dcb1b4e)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14882/head:pull/14882` \
`$ git checkout pull/14882`

Update a local copy of the PR: \
`$ git checkout pull/14882` \
`$ git pull https://git.openjdk.org/jdk.git pull/14882/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14882`

View PR using the GUI difftool: \
`$ git pr show -t 14882`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14882.diff">https://git.openjdk.org/jdk/pull/14882.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14882#issuecomment-1635440073)

### CODE REVIEW

**Comment by AlanBateman** [comment]
The debugger agent in Classic VM was sun.tools.agent.Agent, pre-dates JPDA. I think -Xnoagent was the option to turn it off, like the equivalent of -Xdebugport=-1. Ancient history now, hasn't been needed in 20+ years. I suppose it is baked in some old scripts so printing a warning before its removal is good. We should be able to remove it from our tests too, looks like there are some left overs in test/hotspot/jtreg/vmTestbase/nsk/jdi.

The changes makes me wonder if we should do the same for -Xdebug.

**Comment by jaikiran** [comment]
> We should be able to remove it from our tests too, looks like there are some left overs in test/hotspot/jtreg/vmTestbase/nsk/jdi.

I had noticed those but wasn't sure if they should be included in this change. I'll remove them and run some tests and see how it goes.

> The changes makes me wonder if we should do the same for -Xdebug.

I had a quick look now and it looks like it's just being set and then ignored if `-Xdebug` is passed. There's an accessor method to get the set value but I don't see it being used anywhere. Plus, the comment where `-Xdebug` is handled states:

**Comment by AlanBateman** [comment]
> Would you want me to include the deprecation (for removal) for that option in this same PR or should we do that separately?

It can be looked at separately, my comment was just pointing out another legacy option that probably should have been removed a long time ago. It's probably hard to get data on usages in scripts but my guess is that -Xdebug is in a lot more scripts, even if it hasn't done anything for many years.

**Comment by jaikiran** [comment]
> We should be able to remove it from our tests too, looks like there are some left overs in test/hotspot/jtreg/vmTestbase/nsk/jdi.

I've now updated the PR to remove references to `-Xnoagent` from these tests. 

There's a `TestNullTerminatedFlags` which uses this option. I haven't changed that test given the nature of that test and the fact that this `-Xnoagent` is currently still an accepted command line option.

File: make/ide/netbeans/langtools/build.xml [hunk 1]
```diff
@@ -242,7 +242,7 @@
                 </nbjpdastart>
                 <property
                     name="@{jpda.jvmargs.property}"
-                    value="-Xdebug -Xnoagent -Djava.compiler=none -Xrunjdwp:transport=dt_socket,address=${jpda.address}"
+                    value="-Xdebug -Djava.compiler=none -Xrunjdwp:transport=dt_socket,address=${jpda.address}"
```
Comment (AlanBateman) [inline make/ide/netbeans/langtools/build.xml:245]: Maybe -Djava.compiler=none should be removed too. I don't know if this project file is still used but I assume attempting to debug will print a warning due to this option.

**Review by AlanBateman — APPROVED** [review]

**Comment by plummercj** [comment]
I think `-Xdebug` is also a no-op now.

**Comment by jaikiran** [comment]
Hello Chris,

> I think `-Xdebug` is also a no-op now.

Yes, I'm preparing a separate JBS and PR to address that one.

**Review by dholmes-ora — APPROVED** [review]
This looks fine. What about all the ./src/demo/share/nbproject/jfc/*/nbproject/netbeans-targets.xml files? Will they get cleaned up when this option is actually removed?

Thanks.

**Comment by dholmes-ora** [comment]
>  What about all the ./src/demo/share/nbproject/jfc/*/nbproject/netbeans-targets.xml files? 

Sorry @jaikiran  my repo branch was stale, these were actually cleanup just last week. :)

**Review by plummercj — APPROVED** [review]

**Comment by jaikiran** [comment]
Thank you everyone for the reviews. I think the current Reviews cover the necessary areas of hotspot-runtime and serviceability. So I'll go ahead with the integration later in the day today (3-4 hours from now).

**Pull request closed without merging**